Diff improvements

aantn opened the pull request.

haxsaw reviewed (changes requested):
Hi Natan--

Thanks for this. In general, I like what I see. In particular I've decided I really like the addition of added/removed to the diff types, along with the accompanying logic. I do have a few changes that I'd like to ask for. I wanted to make some comments and the changes will be embedded there.

I see that you've changed DiffDetail from a namedtuple to a dataclass. I guess I don't have any particular issue with that, but doing so is a broader change in that all the doc needs to be reviewed and the necessary changes need to be made in the doc. Additionally, right in the docstring of the diff() method you changed one occurrence of 'namedtuple' to 'object' but left the other one. If you make this kind of change, can I ask you to please find all the references to DiffDetails and ensure that they say the right thing?

I'm a bit unclear as to the value you're trying to accrue by turning diff() into a tiny wrapper for _diff() and making _diff() a classmethod (making paths easier perhaps?); I'd like to know more about that, but that's just curiosity. In your implementation, one case can return arguably inconsistent results:

If you're diff()ing a Hikaru object to another that is None, then in that case the code will issue an INCOMPATIBLE_DIFF warning (which I thought you wanted to get rid of?), and likewise if you diff() say a PodSpec with a PodTemplate, you'll get an incompatible. But it appears that, because you're calling _diff() internally rather than diff(), if there are two different kinds of embedded objects within two Hikaru objects that are the same at the top level, then you'll get different DiffDetails records. For example, suppose we have two Pods, one created incorrectly. One is:

	p1 = Pod(spec=PodSpec(...))  # correct

and the other is

	p2 = Pod(spec=PodTemplateSpec(...))  # incorrect
	
This will pass muster at the diff() level, but within _diff(), it will generate a TYPE_CHANGED. But if you just try diff()ing PodSpec and PodTemplateSpec, you'll get INCOMPATIBLE. That wouldn't have happened if you called diff() internally instead of _diff(). What I don't know is if it is even sensible to report a type changed; the annotations set the types, so what change can be possible that shouldn't be reported as incompatible? I think either you should be calling the diff() method internally so you get consistent results or change the type of diff reported in _diff() if embedded objects are of different types (is there a third option?).

With respect to the messages being generated within _diff(), the use of the word 'Key' where we're really looking at an attribute is misleading. Could you please change these to something like 'Attribute value' instead of 'Key'? 'Key' should only be used if reporting on dict keys.

Thanks for fixing the typos in the existing release notes; reText doesn't spell check, and I hate using PyCharm for editing rst. I can only proofread so many times before I get crosseyed, so it helps when someone else has a look. Can I ask you to author a few sentences for the release_notes for the changes to diff(), and also edit docs/hikaru-base.rst to reflect the changes you made to fields (besides the namedtuple thing).

I think there may be some issue with formatted_attr_path; it looks like the outer diff() function starts it off with the name of the class, but for certain errors within _diff() it gets pre-pended with 'self.', which means you can get paths like 'self.Pod.mumble', which I doubt you mean.

I have mixed feelings about the changes to dict diffing. I like that you fully report on what was added and removed, but I think this change is what has led you to using the term 'Key' when talking about things added or removed. That's an issue doing this with simple recursion-- you don't know the context in which a name has resolved into a value. I have a feeling that if you really want 'key' in the message you'll need to drop the recursion and do the checking here.

If _diff() stands after all of this, could you please change the doc string to a comment block? Private methods are supposed to use comment blocks instead of docstrings (ok, that is a bit nitpicky, but I generally do it elsewhere so I'd like it to be consistent).

Finally, may I ask that you've verified that the attr_path returned in a DiffDetail still resolves to the proper object with 'object_at_path()' ?

Thanks for this; there are a number of aspects I like more than the original implementation, and with some tweaks I think it will be a real improvement.

aantn commented:
Hi,
Thank you for the comments. I'll try to address everything below by category but let me know if I've missed anything:

1. Docs: I've fixed the docs as requested. I'm not sure what command you're using to generate the docs so I didn't actually generate html with my changes. Let me know if I have any mistakes there that need fixing or feel free to fix them yourself. (In general, I'm not attached to my code, so you can modify anything I submit as you see fit. Don't mind at all.)
2. Regarding diff vs _diff: I added _diff because it makes the implementation much simpler. You can't recursively call attr.diff() without first checking attr's type because diff() only exists on subclasses of HikaruBase. For example, "some string".diff() is clearly illegal, but HikaruBase._diff("some string") is legal. So adding _diff() simplifies the implementation and removes duplicate code. (On symptom of this: in your original diff() function you had duplicate code that explicitly checked for int/float classes in two places, but in the new version it appears only once.)
3. TYPE_CHANGED vs INCOMPATIBLE: my rationale was that INCOMPATIBLE is returned when the high-level call that the user made to diff() is illegal because the two objects have a different type. On the other hand, if the internal diff contains a type change then TYPE_CHANGED is returned. As you point out, I'm not sure if there are actual cases where this is legal. Does the swagger file / openapi have a concept of Union types? (By the way, you might want to consider generating pydantic subclasses instead of dataclasses in a future version of hikaru. Pydantic is a python library that enforces type annotations and prevents illegal assignments.) Regardless, I'll update the code to whatever behaviour you like here. I don't think it matters one way or another for my usecase, so let me know how I should handle it.
4. Report messages: I changed the text to "Added:" and "Removed:" as that applies to both attributes and dictionary/list values. Is that OK?

I think fixed the other issues, except for adding a unit test that object_at_path() plays well with DiffDetail.path. I'll do that later today or tomorrow.  Let me know what you think of all the other changes.

On an unrelated note, I think you still have an issue with setup.py(). Also, can you add a doc specifying how you run the unittests? I mucked around with pytest a little, but I had to change PYTHONPATH to get it to work. How do you run it?

aantn commented:
I've added a test for checking object_at_path and the test passes successfully.

On the other hand, the following test (not currently committed) does **not** pass because object_at_path can't traverse dictionary keys.

```
def test122():
    """
    test that you can run object_at_path on the path returned by diff()
    """
    pod = Pod(
        spec=PodSpec(
            containers=[
                Container(name="a", resources=ResourceRequirements(limits={"foo": "bar"}))
            ]
        )
    )
    pod2 = copy.deepcopy(pod)
    pod2.spec.containers[0].resources.limits["foo"] = "blah"
    diff = pod.diff(pod2)
    print(diff)
    print("object at path is", pod.object_at_path(diff[0].path))
```

Lmk how you want to handle this.

haxsaw commented:
Hi--
Thanks for the changes; they look good. 

I understand your point re: diff() vs _diff(); that works well.

I also get what you're saying in regards to incompatible vs type changed. I think I'd prefer consistent error reporting regardless of where the  error occurred if it's the same kind of condition generating the error. I'll deal with that change, so don't trouble yourself there.

I like your solution for the changing the use of 'Key' when an attribute was being discussed; that works for me.

Lastly, I've been thinking about the removal of 'attr' from DiffDetail, and I think I'm going to put it back, but perhaps as a property. I know you don't use it, but there have been a few hundred downloads of Hikaru since I announced it and I don't want to break anyone's code more than necessary. The change of import is minor compared to the number of times someone might look at an attr.

aantn commented:
Awesome, thank you. I'm going to open another ticket re. adding support to object_at_path for dictionary lookups.
